invisible_cities/cities/penthesilea.py

```diff
@@ -85,28 +86,32 @@ def penthesilea(files_in, file_out, compression, event_range, print_mod, detecto
     with tb.open_file(file_out, "w", filters = tbl.filters(compression)) as h5out:
 
         # Define writers...
-        write_mc_             = mc_info_writer(h5out) if run_number <= 0 else (lambda *_: None)
-
-        write_mc              = df.sink(write_mc_                  , args=(        "mc", "event_number"             ))
         write_event_info      = df.sink(run_and_event_writer(h5out), args=("run_number", "event_number", "timestamp"))
         write_hits            = df.sink(         hits_writer(h5out), args="hits")
         write_pointlike_event = df.sink(           kr_writer(h5out), args="pointlike_event")
         write_pmap_filter     = df.sink( event_filter_writer(h5out, "s12_selector"), args=("event_number", "pmap_passed"))
 
-        return push(source = pmap_from_files(files_in),
-                    pipe   = pipe(
-                        df.slice(*event_range, close_all=True),
-                        print_every(print_mod)                ,
-                        event_count_in       .spy             ,
-                        classify_peaks                        ,
-                        pmap_passed                           ,
-                        df.branch(write_pmap_filter)          ,
-                        pmap_select          .filter          ,
-                        event_count_out      .spy             ,
-                        df.fork((build_hits           , write_hits           ),
-                                (build_pointlike_event, write_pointlike_event),
-                                                        write_mc             ,
-                                                        write_event_info     )),
-                    result = dict(events_in  = event_count_in .future,
-                                  events_out = event_count_out.future,
-                                  selection  = pmap_select    .future))
+        event_ids = []
+        events_selected = df.sink(event_ids.append, args=("event_number"))
```

Comment:
Mmm. This is **not** wrong, but we tend to use futures for this kind of things. My approach would be to use the `reduce` component, but I can't think of a good reason not to take this approach. The only (very minor thing) is that using something named `result.events_selected` (or similar) would be slightly more indicative of what is happening, but it's not a big deal at all. I kinda like the fact that you can specify the argument to take :+1:

Reply:
The parenthesis around `"event_number" can be removed.

Reply:
The main reason why I didn't use `reduce` in the end is that it forced me to write something like this:
`add_to_list = df.reduce(lambda l,x: l+[x], initial=[])`
It works, but it looks a bit ugly. I can change it though; it isn't particularly important.

Reply:
No, I like this approach very much!

Reply:
TL:DR the `sink` approach is fine, but just for fun and education ...

> `add_to_list = df.reduce(lambda l,x: l+[x], initial=[])`

... is `O(N)` per insertion (where `N` is the length of the stream) and hence `O(N^2)` in total. You can *almost* do it in `amortized O(1)` per insertion, `O(N)` total with `add_to_list = df.reduce(list.append, initial=[])` except that `list.append` returns `None` rather than the list itself, which is a Python wart, as far as I am concerned, and throws a spanner in the works here.

But you could hack around it with `df.reduce(lambda l,x: l.append(x) or l, initial=[])`, which is a bit ugly, so, you could provide

````python
def append(list_, element):
    list_.append(element)
    return list_
````
which would allow you to say
````python
df.reduce(append, initial=[])
````
cleanly, and keep using the futures pattern that we are used to seeing, and thus address @gonzaponte's comment that

> using something named `result.events_selected` (or similar) would be slightly more indicative of what is happening

Reply:
>  except that list.append returns None rather than the list itself
For that, we had thought of using `np.append` which does return the object, at the expense of making a copy of the array every time.

Reply:
Which makes the time-cost of whole process quadratic rather than linear in the length of the source.

Reply:
Yes, of course. I wasn't suggesting that it was a good choice, but rather an attempt to leave some trace that this was considered :)

Reply:
Yup, good.

Just one more reminder: it's great to theorize, but it's also good to back the theory up with measurement. For which `timeit` (and the way it is embedded in IPython) can be very useful:

````
In [1]: import numpy as np                                                                                                                                                                         

In [2]: from functools import reduce                                                                                                                                                               

In [3]: def a(N): 
   ...:     return reduce(np.append, range(N), np.array([])) 
   ...:

In [4]: def b(N): 
   ...:     return reduce(lambda l,e: l.append(e) or l, range(N), []) 
   ...:

In [5]: assert (a(10**3) == b(10**3)).all()

In [6]: for fn in (a,b): 
   ...:     for p in range(3,6): 
   ...:         %timeit fn(10**p) 
   ...:     print('='*60) 
   ...:

11.9 ms ± 74.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
143 ms ± 1.67 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
4.97 s ± 41.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
============================================================
211 µs ± 12.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.12 ms ± 32.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
21.4 ms ± 155 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
============================================================

In [7]: np.array([143, 4970]) / 11.9
Out[7]: array([ 12.01680672, 417.64705882])

In [8]: np.array([2120, 21400]) / 211
Out[8]: array([  10.04739336, 101.42180095])

````

In summary, the `lambda l,e: l.append(e) or l` approach is significantly faster for nontrivial input size: (211, 2120, 21400) vs (11900, 143000, 4970000); and grows more slowly with increasing size: (1, 10, 111) vs (1, 12, 418).